manifests/module.pp

```diff
@@ -9,75 +9,114 @@
+  }
+
   include ::selinux
 
+  if ($builder == 'simple' and $source_if != undef) {
+    fail("The simple builder does not support the 'source_if' parameter")
+  }
+
+  # let's just make doubly sure that this is an absolute path:
+  validate_absolute_path($::selinux::config::module_build_dir)
+
+  $module_dir = "${::selinux::config::module_build_dir}/${title}"
+  $module_file = "${module_dir}/${title}"
+
+  $build_command = pick($builder, $::selinux::default_builder, 'none') ? {
+      'simple'    => shellquote("${::selinux::config::module_build_dir}/selinux_build_module.sh", $title),
+      'refpolicy' => shellquote('make', '-f', '/usr/share/selinux/devel/Makefile', "${title}.pp"),
+      'none'      => fail('No builder or default builder specified')
+  }
+
   Anchor['selinux::module pre'] ->
   Selinux::Module[$title] ->
   Anchor['selinux::module post']
+  $has_source = (pick($source_te, $source_fc, $source_if, false) != false)
 
-  validate_re($ensure, [ '^present$', '^absent$' ], '$ensure must be "present" or "absent"')
-  if $ensure == 'present' and $source == undef and $content == undef {
-    fail("You must provide 'source' or 'content' field for selinux module")
-  }
-  if $source != undef {
-    validate_string($source)
-  }
-  if $content != undef {
-    validate_string($content)
-  }
-  validate_string($prefix)
-  validate_absolute_path($sx_mod_dir)
-  validate_absolute_path($makefile)
-  if $syncversion != undef {
-    validate_bool($syncversion)
-  }
+  if $has_source and $ensure == 'present' {
+    file {$module_dir:
+      ensure  => directory,
+    }
+
+    if $source_te {
+      file {"${module_file}.te":
+        ensure => 'file',
+        source => $source_te,
+        notify => Exec["clean-module-${title}"],
+      }
+    }
+    if $source_fc {
+      file {"${module_file}.fc":
+        ensure => 'file',
+        source => $source_fc,
+        notify => Exec["clean-module-${title}"],
+      }
+    }
+    if $source_if {
+      file {"${module_file}.if":
+        ensure => 'file',
+        source => $source_if,
+        notify => Exec["clean-module-${title}"],
+      }
+    }
+    exec { "clean-module-${title}":
+      path        => '/bin:/usr/bin',
+      cwd         => $module_dir,
+      command     => "rm -f '${title}.pp' loaded",
+      refreshonly => true,
+      notify      => Exec["build-module-${title}"],
+    }
 
-  ## Begin Configuration
-  file { "${sx_mod_dir}/${prefix}${name}.te":
-    ensure  => $ensure,
-    owner   => 'root',
-    group   => 'root',
-    mode    => '0644',
-    source  => $source,
-    content => $content,
+    exec { "build-module-${title}":
+      path    => '/bin:/usr/bin',
+      cwd     => $module_dir,
+      command => "${build_command} || (rm -f ${title}.pp loaded && exit 1)",
+      creates => "${module_file}.pp",
+      notify  => Exec["install-module-${title}"],
+    }
+    # we need to install the module manually because selmodule is kind of dumb. It ends up
```

Comment:
It's because of the syncversion bug, I think. It can only check that the module is loaded, and if you recompile a module that's loaded, it won't actually detect the change. Installing it manually is a workaround for that.